lib/api-helper/project/fileCopy.ts

```diff
@@ -74,3 +83,34 @@ export function copyFiles(donorProject: Project,
         return p;
     };
 }
+
+/**
+ * Take the specified files from the donor project
+ * @param {RemoteRepoRef} donorProjectId
+ * @param {FileGlobMapping} fileGlobMapping - treated as globs as defined in Project.streamFiles
+ * @param {ProjectOperationCredentials} credentials
+ * @return {SimpleProjectEditor}
+ */
+export function streamFilesFrom(donorProjectId: RemoteRepoRef,
+                                fileGlobMapping: FileGlobMapping,
+                                credentials: ProjectOperationCredentials): CodeTransform {
+    return async (p, i) => {
+        const donorProject = await GitCommandGitProject.cloned(credentials, donorProjectId);
+        return streamFiles(donorProject, fileGlobMapping)(p, i);
+    };
+}
+
+export function streamFiles(donorProject: Project,
+                            fileGlobMapping: FileGlobMapping): CodeTransform {
+    return async p => {
+        await donorProject.streamFiles(...fileGlobMapping.globPatterns)
+            .on("end", () => { logger.debug("end of file stream reached, using glob: ", fileGlobMapping); })
+            .on("data", f => {
+                (async () => {
+                    await p.addFile((fileGlobMapping.recipientPath || "") + f.path, await f.getContent());
+                    logger.debug("file copied: ", f);
+                })();
+            })
+        ;
+    };
```

Comment:
I think this should be something like:

```typescript
return new Promise((resolve, reject) => {
    const fileStream = donorProject.streamFiles(...fileGlobMapping.globPatterns);
    fileStream.on("end", resolve);
    fileStream.on("error", reject);
    fileStream.on("data", f => {
        const targetPath = (fileGlobMapping.recipientPath || "") + f.path;
        fileStream.pause();
        try {
            const content = f.getContentSync();
            p.addFileSync(targetPath, content);
        } catch (e) {
            e.message = `Failed to copy ${f.path} to ${p.name}:${targetPath}: ${e.message}`;
            logger.error(e.message);
            fileStream.destroy(e);
        }
        logger.debug(`File copied: ${f.path} => ${targetPath}`);
        fileStream.resume();
    });
});
```

I think you were right about the async stuff within the stream callbacks, https://stackoverflow.com/questions/36662113/node-js-can-you-use-asynchronous-functions-from-within-streams.

Reply:
As I said, I am new to Javascript and typescript. But I do not think all this is necessary. I suspect that the stream sets up a race condition, but I am not sure. Will test with some real-life cases first.
I am not sure if it is guaranteed that all files are copied when the methods file stream calls are done or if it sets up a background stream? 

The fileStream resume/destroy/pause methods are not publically available in the FileStream object, to my knowledge at least.